### ISSUE TITLE
Bugs due to two selectors used that are not available on older Mac OS X versions (e.g. 10.8)

### ISSUE THREAD

**Issue by follower** [MtnViewJohn/context-free#17]
### Cause

The use of `finishWritingWithCompletionHandler` here:

https://github.com/MtnViewJohn/context-free/blob/97b4074601a6726abd52340f27aff16f9ad3fa05/src-osx/AVcanvas.mm#L214

prevents animation output from completing because the application crashes with an exception:

```
[AVAssetWriter finishWritingWithCompletionHandler:]: unrecognized selector sent to instance
``` 

### Effect

It is not possible to export animations (and the app crashes). :(

### Potential solution

According to the [`finishWritingWithCompletionHandler` documentation](https://developer.apple.com/documentation/avfoundation/avassetwriter/1390432-finishwritingwithcompletionhandl) the method was added in Mac OS X 10.9. It replaced the now deprecated [`finishWriting` method](https://developer.apple.com/documentation/avfoundation/avassetwriter/1426644-finishwriting?language=objc).

Perhaps the deprecated method could be used (at least on < 10.9) instead.

### Cause

If invalid animation settings are provided in the "Start Animation" or "Animate a Frame" dialogs then clicking the "Animate" button generates an exception:

```
NSAlert beginSheetModalForWindow:completionHandler:]: unrecognized selector sent to instance
```
The exception is caused by the use of `beginSheetModalForWindow` here:

https://github.com/MtnViewJohn/context-free/blob/1c17dc46868c0741f73b081f384a0bc66bc583b9/src-osx/CFDGDocument.mm#L647-L648

### Effect

After the exception occurs the "Animate" button appears non-functional but no feedback is provided that the settings are invalid.

### Workaround

In my case I eventually located the following validation code and manually determined (as I recall) the issue was the image dimensions were not multiples of eight:

https://github.com/MtnViewJohn/context-free/blob/97b4074601a6726abd52340f27aff16f9ad3fa05/src-osx/CFDGDocument.mm#L618-L639

So, while not exactly friendly it doesn't prevent successful operation (unlike the first issue).

### Potential solution

The [`beginSheetModalForWindow:completionHandler:` documentation](https://developer.apple.com/documentation/appkit/nsalert/1524296-beginsheetmodalforwindow) states the method was new in Mac OS X 10.9.

For a potential solution to support older versions see:

 * "[How can I use [NSAlert beginSheetModalForWindow:completionhandler:] on older versions of OS X](https://stackoverflow.com/questions/20146249/how-can-i-use-nsalert-beginsheetmodalforwindowcompletionhandler-on-older-ver)"

 * "[beginSheetModalForWindow: completionHandler:](https://stackoverflow.com/questions/28927678/beginsheetmodalforwindow-completionhandler)"

[Note: This issue report is only a stub but hopefully it's better than no issue. :) I'm hoping to add more detail.]

**Comment by MtnViewJohn** [comment]
I will use the deprecated methods instead in the next release.

**Comment by follower** [comment]
Thanks. Appreciate you being open to supporting older systems.

Happy to test a pre-release build if that's helpful.

**Comment by MtnViewJohn** [comment]
That would be great. Download it from [here](https://www.ozonehouse.com/john/ContextFree3.1.dmg).

**Comment by follower** [comment]
Thanks for pre-release build to test. Currently taking a look at it.

I've managed to successfully generate an animation which is good, thanks.

There appears to be a couple of issues I encountered, one of which is:
 * Opening the "File" menu now generates an uncaught exception error (it didn't previously):
```
-[GView validateMenuItem:]: unrecognized selector sent to instance 0x7fd98bb9ad60
```
  * I think it's due to this code:

https://github.com/MtnViewJohn/context-free/blob/843ce8ba56f73194e3c3c23c0d409cf865b7d626/src-osx/CFDGDocument.mm#L531-L552

**Comment by follower** [comment]
Other issues I've encountered (unrelated to original issue, aside from being encountered now that animation is possible):

* Animation player controller appears to have second & third controls overlapping:

![animation-playback-control-bugs-screenshot](https://user-images.githubusercontent.com/189962/39796318-f9a2f86a-53a9-11e8-8172-7a02dfb355d5.png)

* The "Current Playback Position" field in the middle isn't updated to zero when animation is regenerated (see above value of `1.0` when player head is at `0`). (FWIW--particularly with this bug--but even in general--IMO it's not entirely obvious the middle value is the current playback position.)

**Comment by follower** [comment]
Also:

 * `Render > Animate...` followed by `Render > Animate a frame...` doesn't display the selected rendered frame unless you run  `Render > Save Output...` in-between. (Seems like the "render frame" method may not hide the "animation sheet"?)

**Comment by follower** [comment]
I'm slightly confused about behaviour which I thought was a bug but appears to be the same in the previous release:

 * What is the intended purpose of the `Render > Repeat Render` menu item?

I had thought that repeating the render used the same seed/variation value again without increment but that doesn't seem to be the case.

Further investigation suggests the effect is related to editing of the list of rules:

 1. Create New file. (Variation value is prefixed with "next variation")
 2. Paste this: (Prefix: "next variation")
```
startshape START

shape START {
SQUARE []
}
```
 3. `Render > Render` (Prefix changes to: "this variation"; seed does not increment)
 4. `Render > Repeat Render` (Prefix is unchanged "this variation"; seed *does* increment.)
 5. `Render > Render` (Prefix is unchanged "this variation"; seed *does* increment.)
 6. Insert space at on `startshape` line. (Prefix changes to: "next variation"; seed does not increment)
 7. `Render > Repeat Render`  (Prefix changes to: "this variation"; seed does not increment)
 8. `Render > Repeat Render` (Prefix is unchanged "this variation"; seed *does* increment.)

So, it seems currently "Repeat Render" only "repeats" the render with the same seed *if the grammar has also been edited*, otherwise it increments the seed and renders as "Render" does.

I *think* the assumption is that "Repeat Render" is only used on an edited grammar? Which makes sense on one level but also leads to inconsistent behaviour if (for whatever reason--I can't remember why I encountered this originally) "Repeat Render" is used...repeatedly. :)  )

I would suggest "Repeat Render" never increments the seed as this behaviour seems more consistent and to me clearly differentiates between the two menu items.

[Thanks again for the test build to try, I look forward to playing more with the animation functionality.]

**Comment by MtnViewJohn** [comment]
I have updated the dmg to fix your issues.

1) File menu crash: I fixed CFDGDocument to call GView validateUserInterfaceItem: instead of the now missing GView validateMenuItem: method. Interestingly, this crash does not happen on High Sierra even though the selector is not present in GView.

2) I don't know what is going on with the rewind button. It draws fine on Sierra/High Sierra. I can reproduce the  issue by booting into Lion. It doesn't seem to affect functionality.

3) The playback position field resets to 0.0 like it should and now stands out more.

4) Animate-a-frame displays the frame even after a movie was playing

5) Repeat Render was added to make it easier to explore variations when using Render with dimensions. It allows you to sequentially see what variations of a design look like when rendered to a specific size without tediously bringing up the Render with dimensions modal sheet and clicking OK.  

I think a better approach would be to get rid if it and get rid of the modal Render with dimensions sheet. Instead, the Render button should have a Render with dimensions mode and the render size controls should be available in a non-modal manner.
